App/Logic/DataUploadLogic.swift

```diff
@@ -232,22 +266,181 @@ extension Logic.DataUpload {
   }
 }
 
+// MARK: - Dummy traffic
+
+extension Logic.DataUpload {
+  /// Updates the dummy ingestion traffic opportunity window taking the parameters from the Configuration and the RNGs.
+  struct UpdateDummyTrafficOpportunityWindowIfExpired: AppSideEffect {
+    func sideEffect(_ context: SideEffectContext<AppState, AppDependencies>) throws {
+      let state = context.getState()
+
+      let now = context.dependencies.now()
+
+      guard now > state.ingestion.dummyTrafficOpportunityWindow.windowEnd else {
+        // Window not started or finished yet.
+        return
+      }
+
+      try context.awaitDispatch(UpdateDummyTrafficOpportunityWindow())
+    }
+  }
+
+  /// Updates the dummy ingestion traffic opportunity window taking the parameters from the Configuration and the RNGs.
+  struct UpdateDummyTrafficOpportunityWindow: AppSideEffect {
+    func sideEffect(_ context: SideEffectContext<AppState, AppDependencies>) throws {
+      let state = context.getState()
+
+      let dummyTrafficStochasticDelay = context.dependencies.exponentialDistributionGenerator
+        .exponentialRandom(with: state.configuration.dummyIngestionMeanStochasticDelay)
+
+      try context.awaitDispatch(
+        SetDummyTrafficOpportunityWindow(
+          dummyTrafficStochasticDelay: dummyTrafficStochasticDelay,
+          dummyTrafficOpportunityWindowDuration: state.configuration.dummyIngestionWindowDuration,
+          now: context.dependencies.now()
+        )
+      )
+    }
+  }
+
+  /// Schedules a sequence of dummy ingestion requests for some random point in the future.
+  /// -seeAlso: https://github.com/immuni-app/immuni-documentation/blob/master/Traffic%20Analysis%20Mitigation.md
+  struct ScheduleDummyIngestionSequenceIfNecessary: AppSideEffect {
+    func sideEffect(_ context: SideEffectContext<AppState, AppDependencies>) throws {
+      let state = context.getState()
+
+      guard !state.ingestion.dummyTrafficSequenceScheduledInSession else {
```

Comment:
How's possible that there is another sequence scheduled?
start + will enter foreground?

Reply:
Exactly